istio.io/istio/pkg/probe failures (common test failure)

```
go test  ./pkg/...
ok  	istio.io/istio/pkg/cache	1.196s
?   	istio.io/istio/pkg/collateral	[no test files]
ok  	istio.io/istio/pkg/log	0.066s
--- FAIL: TestFileControllerMethods (0.30s)
	controller_test.go:237: Got file /var/folders/c3/4bc8zz5s5hj13nnr718w91xr00f56m/T/TestFileControllerMethods257307300/fc is too old (last modified time 2018-02-01 15:10:30 -0800 PST, should be within 110ms), Want nil
	controller_test.go:245: Got file /var/folders/c3/4bc8zz5s5hj13nnr718w91xr00f56m/T/TestFileControllerMethods257307300/fc is too old (last modified time 2018-02-01 15:10:30 -0800 PST, should be within 110ms), Want nil
--- FAIL: TestFileController (0.90s)
	controller_test.go:275: Got file /var/folders/c3/4bc8zz5s5hj13nnr718w91xr00f56m/T/TestFileController517214131/fc is too old (last modified time 2018-02-01 15:10:30 -0800 PST, should be within 110ms), want nil
	controller_test.go:288: Got file /var/folders/c3/4bc8zz5s5hj13nnr718w91xr00f56m/T/TestFileController517214131/fc is too old (last modified time 2018-02-01 15:10:30 -0800 PST, should be within 110ms), want nil
	controller_test.go:293: Got file /var/folders/c3/4bc8zz5s5hj13nnr718w91xr00f56m/T/TestFileController517214131/fc is too old (last modified time 2018-02-01 15:10:31 -0800 PST, should be within 110ms), want nil
FAIL
FAIL	istio.io/istio/pkg/probe	1.363s
ok  	istio.io/istio/pkg/tracing	0.056s
?   	istio.io/istio/pkg/util	[no test files]
ok  	istio.io/istio/pkg/version	0.043s
```

this is on my mac, circleci doesn't seem to run this test (#3081 ) so I can't say if it's specific to mac

also failed on prow:

https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3119/istio-presubmit/4036/
```
I0202 22:09:18.814] --- FAIL: TestController (0.04s)
I0202 22:09:18.814] 	controller_test.go:118: Got 1 error occurred:
I0202 22:09:18.815] 		
I0202 22:09:18.815] 		* p1: uninitialized, want nil
I0202 22:09:18.815] FAIL
I0202 22:09:18.815] FAIL	istio.io/istio/pkg/probe	1.349s
I0202 22:09:18.815] ok  	istio.io/istio/pkg/tracing	0.069s
I0202 22:09:18.816] ?   	istio.io/istio/pkg/util	[no test files]
I0202 22:09:18.816] ok  	istio.io/istio/pkg/version	0.043s
I0202 22:09:18.822] Makefile:363: recipe for target 'common-test' failed
W0202 22:09:18.823] make: *** [common-test] Error 1
E0202 22:09:18.823] Build failed
I0202 22:09:18.823] process 520 exited with code 1 after 15.4m
E0202 22:09:18.823] FAIL: istio-presubmit
```

```
I0314 16:20:55.704] --- FAIL: TestFileControllerMethods (0.58s)
I0314 16:20:55.704]     controller_test.go:245: Got file /tmp/TestFileControllerMethods432811897/fc is too old (last modified time 2018-03-14 16:20:47.244177734 +0000 UTC, should be within 110ms), Want nil
```
happens on prow too
https://k8s-gubernator.appspot.com/build/istio-prow/istio-postsubmit/1187?log#log

--- a/pkg/probe/client.go
+++ b/pkg/probe/client.go
@@ -26,17 +26,18 @@ type Client interface {
 }
 
 type fileClient struct {
-	opt *Options
+	opt      *Options
+	statFunc func(path string) (os.FileInfo, error)
 }
 
 // NewFileClient creates an instance of Client based on the file status specified
 // in the path. The specified period is the interval of the probe, so if this
 func NewFileClient(opt *Options) Client {
-	return &fileClient{opt}
+	return &fileClient{opt: opt, statFunc: os.Stat}
 }
 
 func (fc *fileClient) GetStatus() error {
-	stat, err := os.Stat(fc.opt.Path)
+	stat, err := fc.statFunc(fc.opt.Path)
 	if err != nil {
 		return err
 	}

--- a/pkg/probe/client_test.go
+++ b/pkg/probe/client_test.go
@@ -1,0 +1,93 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package probe
+
+import (
+	"errors"
+	"os"
+	"testing"
+	"time"
+)
+
+// dummyFileInfo is a dummy object to represent certain timestamp.
+type dummyFileInfo struct {
+	modTime     time.Time
+	calledCount int
+}
+
+func (d *dummyFileInfo) Name() string {
+	return "dummy"
+}
+
+func (d *dummyFileInfo) Size() int64 {
+	return 0
+}
+
+func (d *dummyFileInfo) Mode() os.FileMode {
+	return 0750
+}
+
+func (d *dummyFileInfo) ModTime() time.Time {
+	return d.modTime
+}
+
+func (d *dummyFileInfo) IsDir() bool {
+	return false
+}
+
+func (d *dummyFileInfo) Sys() interface{} {
+	return nil
+}
+
+func TestFileClient(t *testing.T) {
+	d := &dummyFileInfo{}
+	var statErr error
+	fc := &fileClient{
+		opt: &Options{
+			Path:           "test",
+			UpdateInterval: time.Minute,
+		},
+		statFunc: func(path string) (os.FileInfo, error) {
+			if statErr != nil {
+				return nil, statErr
+			}
+			d.calledCount++
+			return d, nil
+		},
+	}
+
+	statErr = errors.New("dummy")
+	if err := fc.GetStatus(); err != statErr {
+		t.Errorf("Want %v, Got %v", statErr, err)
+	}
+
+	statErr = nil
+	d.modTime = time.Now()
+	if err := fc.GetStatus(); err != nil {
+		t.Errorf("Want nil, Got %v", err)
+	}
+	if d.calledCount != 1 {
+		t.Errorf("Want 1, Got %d", d.calledCount)
+	}
+	d.calledCount = 0
+
+	d.modTime = time.Now().Add(-time.Hour)
+	if err := fc.GetStatus(); err == nil {
+		t.Error("Want error, Got nil")
+	}
+	if d.calledCount != 1 {
+		t.Errorf("Want 1, Got %d", d.calledCount)
+	}
+}

--- a/pkg/probe/controller_test.go
+++ b/pkg/probe/controller_test.go
@@ -24,10 +24,7 @@ import (
 	"time"
 )
 
-const (
-	testDuration     = 10 * time.Millisecond
-	testFileDuration = 100 * time.Millisecond
-)
+const testDuration = 10 * time.Millisecond
 
 var errClosed = errors.New("closed")
 
@@ -36,6 +33,7 @@ type dummyImpl struct {
 	lastStatus  error
 	changeCount int
 	updatec     chan struct{}
+	closec      chan struct{}
 }
 
 func (d *dummyImpl) GetStatus() error {
@@ -54,6 +52,7 @@ func (d *dummyImpl) onClose() error {
 	d.Lock()
 	defer d.Unlock()
 	d.lastStatus = errClosed
+	close(d.closec)
 	return nil
 }
 
@@ -62,34 +61,22 @@ func (d *dummyImpl) onUpdate(newStatus error) {
 	defer d.Unlock()
 	d.lastStatus = newStatus
 	d.changeCount++
-	if d.updatec != nil {
-		select {
-		case d.updatec <- struct{}{}:
-		default:
-		}
+	select {
+	case d.updatec <- struct{}{}:
+	default:
 	}
 }
 
-func (d *dummyImpl) wait(t *testing.T, timeout time.Duration) {
-	t.Helper()
+func (d *dummyImpl) wait() {
+	<-d.updatec
 	d.Lock()
-	if d.updatec != nil {
-		close(d.updatec)
-	}
+	defer d.Unlock()
+	close(d.updatec)
 	d.updatec = make(chan struct{})
-	d.Unlock()
-	for i := 0; i < 10; i++ {
-		select {
-		case <-d.updatec:
-			return
-		case <-time.After(timeout):
-		}
-	}
-	t.Fatal("Failed to wait")
 }
 
 func newDummyController() (Controller, *dummyImpl) {
-	d := &dummyImpl{}
+	d := &dummyImpl{updatec: make(chan struct{}), closec: make(chan struct{})}
 	c := &controller{
 		statuses: map[*Probe]error{},
 		name:     "dummy",
@@ -101,20 +88,19 @@ func newDummyController() (Controller, *dummyImpl) {
 }
 
 func TestController(t *testing.T) {
-	t.Skip("issue https://github.com/istio/istio/issues/3082")
 	c, d := newDummyController()
 	defer c.Close()
 
 	p1 := NewProbe()
 	p1.RegisterProbe(c, "p1")
-	time.Sleep(testDuration)
+	d.wait()
 	if err := d.GetStatus(); err == nil {
 		t.Error("Got nil, want error")
 	}
 
 	prevCount := d.GetCount()
 	p1.SetAvailable(nil)
-	d.wait(t, testDuration)
+	d.wait()
 	if err := d.GetStatus(); err != nil {
 		t.Errorf("Got %v, want nil", err)
 	}
@@ -124,19 +110,19 @@ func TestController(t *testing.T) {
 
 	p2 := NewProbe()
 	p2.RegisterProbe(c, "p2")
-	d.wait(t, testDuration)
+	d.wait()
 	if err := d.GetStatus(); err == nil {
 		t.Error("Got nil, want error")
 	}
 
 	p2.SetAvailable(nil)
-	d.wait(t, testDuration)
+	d.wait()
 	if err := d.GetStatus(); err != nil {
 		t.Errorf("Got %v, want nil", err)
 	}
 	prevCount = d.GetCount()
 
-	d.wait(t, testDuration*2)
+	d.wait()
 	if err := d.GetStatus(); err != nil {
 		t.Errorf("Got %v, want nil", err)
 	}
@@ -145,13 +131,13 @@ func TestController(t *testing.T) {
 	}
 
 	p1.SetAvailable(errors.New("dummy"))
-	d.wait(t, testDuration)
+	d.wait()
 	if err := d.GetStatus(); err == nil {
 		t.Error("Got nil, want error")
 	}
 
 	prevCount = d.GetCount()
-	d.wait(t, testDuration*2)
+	d.wait()
 	if err := d.GetStatus(); err == nil {
 		t.Error("Got nil, want error")
 	}
@@ -186,18 +172,17 @@ func TestControllerRegisterTwice(t *testing.T) {
 }
 
 func TestControllerAfterClose(t *testing.T) {
-	t.Skip("issue https://github.com/istio/istio/issues/3082")
 	c, d := newDummyController()
 
 	p1 := NewProbe()
 	p1.RegisterProbe(c, "p1")
-	d.wait(t, testDuration)
+	d.wait()
 	if err := d.GetStatus(); err == nil {
 		t.Error("Got nil, want error")
 	}
 
 	p1.SetAvailable(nil)
-	d.wait(t, testDuration)
+	d.wait()
 	if err := d.GetStatus(); err != nil {
 		t.Errorf("Got %s, want nil", err)
 	}
@@ -205,7 +190,7 @@ func TestControllerAfterClose(t *testing.T) {
 	if err := c.Close(); err != nil {
 		t.Errorf("failed to close: %v", err)
 	}
-	time.Sleep(testDuration * 2)
+	<-d.closec
 	if err := d.GetStatus(); err != errClosed {
 		t.Errorf("Got %v, Want %v", err, errClosed)
 	}
@@ -222,8 +207,24 @@ func TestControllerAfterClose(t *testing.T) {
 	}
 }
 
-func TestFileControllerMethods(t *testing.T) {
-	t.Skip("issue https://github.com/istio/istio/issues/3082")
+func TestNewFileController(t *testing.T) {
+	d, err := ioutil.TempDir("", t.Name())
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(d)
+	path := filepath.Join(d, "fc")
+	c := NewFileController(&Options{Path: path, UpdateInterval: testDuration})
+	fc, ok := c.(*controller).impl.(*fileController)
+	if !ok || fc == nil {
+		t.Fatalf("NewFileController should return with fileController: %+v", fc)
+	}
+	if fc.path != path {
+		t.Errorf("Want %s, Got %s", path, fc.path)
+	}
+}
+
+func TestFileController(t *testing.T) {
 	d, err := ioutil.TempDir("", t.Name())
 	if err != nil {
 		t.Fatal(err)
@@ -231,80 +232,33 @@ func TestFileControllerMethods(t *testing.T) {
 	defer os.RemoveAll(d)
 	path := filepath.Join(d, "fc")
 	fc := &fileController{path: path}
-	client := NewFileClient(&Options{path, testFileDuration})
-	if err := client.GetStatus(); err == nil {
-		t.Error("Got nil, Want error")
+	if _, err := os.Stat(path); !os.IsNotExist(err) {
+		t.Errorf("Got %v, Want not-existed", err)
 	}
 	fc.onUpdate(nil)
-	if err := client.GetStatus(); err != nil {
-		t.Errorf("Got %v, Want nil", err)
-	}
-	time.Sleep(testFileDuration * 3)
-	if err := client.GetStatus(); err == nil {
-		t.Error("Got nil, Want error")
-	}
-	fc.onUpdate(nil)
-	if err := client.GetStatus(); err != nil {
+	if _, err := os.Stat(path); err != nil {
 		t.Errorf("Got %v, Want nil", err)
 	}
 	fc.onUpdate(errors.New("dummy"))
-	if err := client.GetStatus(); !os.IsNotExist(err) {
+	if _, err := os.Stat(path); !os.IsNotExist(err) {
 		t.Errorf("Got %v, Want not-existed", err)
 	}
-}
-
-func TestFileController(t *testing.T) {
-	t.Skip("issue https://github.com/istio/istio/issues/3082")
-	d, err := ioutil.TempDir("", t.Name())
-	if err != nil {
-		t.Fatal(err)
+	fc.onUpdate(nil)
+	if _, err := os.Stat(path); err != nil {
+		t.Errorf("Got %v, Want nil", err)
 	}
-	defer os.RemoveAll(d)
-	opt := &Options{filepath.Join(d, "fc"), testFileDuration}
-	fc := NewFileController(opt)
-	client := NewFileClient(opt)
-	fc.Start()
-	defer fc.Close()
-
-	p1 := NewProbe()
-	p1.RegisterProbe(fc, "p1")
-	time.Sleep(testFileDuration)
-	if err := client.GetStatus(); err == nil {
-		t.Errorf("Got nil, want error")
+	fc.onUpdate(nil)
+	if _, err := os.Stat(path); err != nil {
+		t.Errorf("Got %v, Want nil", err)
 	}
 
-	p1.SetAvailable(nil)
-	time.Sleep(testFileDuration)
-	if err := client.GetStatus(); err != nil {
-		t.Errorf("Got %v, want nil", err)
+	if err := fc.onClose(); err != nil {
+		t.Errorf("Got %v, Want nil", err)
 	}
-
-	p2 := NewProbe()
-	p2.RegisterProbe(fc, "p2")
-	time.Sleep(testFileDuration)
-	if err := client.GetStatus(); err == nil {
-		t.Errorf("Got nil, want error")
+	if _, err := os.Stat(path); !os.IsNotExist(err) {
+		t.Errorf("Got %v, Want not-existed", err)
 	}
-
-	p2.SetAvailable(nil)
-	time.Sleep(testFileDuration)
-	if err := client.GetStatus(); err != nil {
-		t.Errorf("Got %v, want nil", err)
-	}
-
-	time.Sleep(testFileDuration * 2)
-	if err := client.GetStatus(); err != nil {
-		t.Errorf("Got %v, want nil", err)
-	}
-
-	p1.SetAvailable(errors.New("dummy"))
-	time.Sleep(testFileDuration)
-	if err := client.GetStatus(); err == nil {
-		t.Error("Got nil, want error")
-	}
-
-	time.Sleep(testFileDuration * 2)
-	if err := client.GetStatus(); err == nil {
-		t.Error("Got nil, want error")
+	if err := fc.onClose(); err != nil {
+		t.Errorf("Got %v, Want nil", err)
 	}
 }

--- a/pkg/probe/option_test.go
+++ b/pkg/probe/option_test.go
@@ -1,0 +1,43 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package probe
+
+import (
+	"testing"
+	"time"
+)
+
+func TestOption(t *testing.T) {
+	var o *Options
+	if o.IsValid() {
+		t.Error("nil should not be valid")
+	}
+	o = &Options{}
+	if o.IsValid() {
+		t.Errorf("Empty option %+v should not be valid", o)
+	}
+	o.Path = "foo"
+	if o.IsValid() {
+		t.Errorf("%+v should not be valid since interval is missing", o)
+	}
+	o.UpdateInterval = time.Second
+	if !o.IsValid() {
+		t.Errorf("%+v should be valid", o)
+	}
+	o.Path = ""
+	if o.IsValid() {
+		t.Errorf("%+v should not be valid since path is missing", o)
+	}
+}

